### PR TITLE
early migrations

### PR DESCRIPTION
**Pull request by Emily**
@kesne this is wip still, but the general syntax is largely what I was going for.

### CODE REVIEW

**Comment by Emily** [comment]
@kesne pushed up adapter refactor of it -- still need to set up integration tests, probably tomorrow night.

File: packages/fewer/src/Migration/index.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+interface IrreversibleMigrationDefinition<T> {
+  up: (m: T) => void;
+  irreversible: true;
+}
+
+export type MigrationDefinition<T> = ChangeMigrationDefinition<T> | UpDownMigrationDefinition<T> | IrreversibleMigrationDefinition<T>;
+
+export function isChangeMigration<T>(definition: MigrationDefinition<T>): definition is ChangeMigrationDefinition<T> {
```
Comment (kesne) [inline packages/fewer/src/Migration/index.ts]: You could probably avoid these functions if you put a static `type` field on all of the migrations and set it to some enum value.

File: packages/fewer/src/Database/Adapter.ts [hunk 2]
```diff
@@ -17,4 +17,6 @@ export default interface Adapter {
    * Updates a record in the database. Returns the id of the updated item.
    */
   update(query: Update): Promise<any>;
+
+  rawQuery(query: String): Promise<any>;
```
Comment (kesne) [inline packages/fewer/src/Database/Adapter.ts]: Don't put this on the base adapter.

File: packages/fewer/src/Database/index.ts [hunk 3]
```diff
@@ -32,6 +32,10 @@ export class Database {
   update(query: Update): Promise<any> {
     return this.adapter.update(query);
   }
+
+  rawQuery(query: string): Promise<any> {
```
Comment (kesne) [inline packages/fewer/src/Database/index.ts]: Don't put this on the base db

**Comment by Emily** [comment]
@kesne yes, to all the things, but as discussed going to merge this and work on those other things, since you're switching up the schema/migration syntax anyway.

**Pull request merged**